pkg/apis/core/types.go

```diff
@@ -544,6 +553,23 @@ const (
 	PersistentVolumeClaimNodeResizeFailed ClaimResourceStatus = "NodeResizeFailed"
 )
 
+// +enum
+type VolumeAttributesClassStatus string
+
+const (
+	// When volume modification is complete, the empty string is set by modify volume controller.
+	PersistentVolumeClaimNoModifyVolumeInProgress VolumeAttributesClassStatus = ""
+	// State set when modify volume controller finds that volumeAttributesClassName is specified
+	// but the corresponding VolumeAttributesClass is not found.
+	PersistentVolumeClaimControllerModifyVolumePending VolumeAttributesClassStatus = "ControllerModifyVolumePending"
```

Comment:
The [Kubernetes API Changes in Proposal](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3751-volume-attributes-class#kubernetes-api) is consistent with the [Proposed Changes](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3751-volume-attributes-class#3-add-new-statuses-in-pvc-api-to-indicate-changes-of-volumeattributesclass-and-the-status-of-the-modifyvolume-operation) on the new statuses in PVC API.

/cc @sunnylovestiramisu

Reply:
empty string means completed. So we cannot use it for both.

Reply:
Looking at controller resize, there is no "Pending" status: https://github.com/kubernetes/kubernetes/blob/246aba89128ed9f3236321baed8f0778cc8800cb/pkg/apis/core/types.go#L607-L611

The reason is because there is going to be a timing delay for a controller to send the Pending status, so it's going to be inconsistent for users already. Specifically, this sequence of events:
1. PVC.status is empty
2. User changes the VAC
3. Controller sees the update and sets status to Pending.
4. Controller starts processsing and sets the status again to Inprogress

In between steps 2 and 3, if you will see the status as "" even though the VACs are different. So this is already inconsistent.

Between steps 3 and 4, the controller is not doing anything signficant, so the time the PVC spends in Pending is insignificant and users will be unlikely to observe the PVC in that state to make a meaningful difference in handling. It also requires an additional API call by the controller to update the state twice in a row.

Reply:
I am confused about the API changes on PVC after reading the KEP. FYI:

- API changes on PVC from the `Proposal` section of the KEP.

```go
type PersistentVolumeClaimModifyVolumeStatus string

const (
  // When modify volume is complete, the empty string is set by modify volume controller.
  PersistentVolumeClaimNoModifyVolumeInProgress PersistentVolumeClaimModifyVolumeStatus = ""
  // State set when modify volume controller starts modifying the volume in control-plane
  PersistentVolumeClaimControllerModifyVolumeInProgress PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeInProgress"
  // State set when modify volume has failed in modify volume controller with a terminal error.
  // Transient errors such as timeout should not set this status and should leave ModifyVolumeStatus
  // unmodified, so as modify volume controller can resume the volume modification.
  PersistentVolumeClaimControllerModifyVolumeFailed PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeFailed"
)
```

- API changes on PVC from the `Proposed Changes` section of the KEP.

```go
type VolumeAttributesClassStatus string

const (
    PersistentVolumeClaimControllerModifyVolumeProgress VolumeAttributesClassStatus = "ControllerModifyVolumeInProgress"
    PersistentVolumeClaimControllerModifyVolumePending VolumeAttributesClassStatus = "ControllerModifyVolumePending"
    PersistentVolumeClaimControllerModifyVolumeFailed VolumeAttributesClassStatus = "ControllerModifyVolumeFailed"
)
```

Which one is the correct one? The `Pending` status is missing in the `Proposal` section. Should the API changes on PVC contain the `Pending` status?

Which component should set the `Pending` status? The controller (in reconcile phase) or the apiserver (in PrepareUpdate phase)? 

Why we need the `Pending` status? Is it used to indicate the volume is being modified but the modification is not started or the corresponding volume attributes class is not found when the controller is trying to obtain the volume attributes class?

Reply:
To be consistent with how we report conditions for ControllerResize, I think we can remove "Pending" entirely and just have "", "InProgress" and "Failed"

Reply:
@msau42 It is pointer in the status. we can use nil value to  indicate no modification or operation completed. we can also remove "". what do you think?

Reply:
Pending is used to when we want to create a PVC, but the VAC does not exist yet, we discussed that we do not want ordering of object creation?